include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -1198,6 +1263,113 @@ __parallel_copy_if(oneapi::dpl::__internal::__device_backend_tag __backend_tag,
     }
 }
 
+template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3, typename _Compare,
+          typename _IsOpDifference>
+auto
+__parallel_set_reduce_then_scan(oneapi::dpl::__internal::__device_backend_tag __backend_tag, _ExecutionPolicy&& __exec,
+                                _Range1&& __rng1, _Range2&& __rng2, _Range3&& __result, _Compare __comp,
+                                _IsOpDifference)
+{
+    // fill in reduce then scan impl
+    using _GenMaskReduce = oneapi::dpl::__par_backend_hetero::__gen_set_mask<_IsOpDifference, _Compare>;
+    using _GenMaskScan =
+        oneapi::dpl::__par_backend_hetero::__gen_mask<oneapi::dpl::__internal::__no_op,
+                                                      oneapi::dpl::__par_backend_hetero::__extract_range_from_zip<2>>;
+    using _WriteOp = oneapi::dpl::__par_backend_hetero::__write_to_id_if<0, oneapi::dpl::__internal::__pstl_assign>;
+    using _Size = oneapi::dpl::__internal::__difference_t<_Range3>;
+
+    using _GenReduceInput = oneapi::dpl::__par_backend_hetero::__gen_count_mask<_GenMaskReduce>;
+    using _ReduceOp = std::plus<_Size>;
+    using _GenScanInput = oneapi::dpl::__par_backend_hetero::__gen_expand_count_mask<
+        _GenMaskScan, oneapi::dpl::__par_backend_hetero::__extract_range_from_zip<0>>;
+    using _ScanInputTransform = oneapi::dpl::__par_backend_hetero::__get_zeroth_element;
+
+    oneapi::dpl::__par_backend_hetero::__buffer<_ExecutionPolicy, int32_t> __mask_buf(__exec, __rng1.size());
```

Comment:
I have not tried this.  I believe the only place we do this currently is `__result_and_scratch_storage`.  I could try to see the impact we would stand to gain, but we would need to create some infrastructure around it similar to that which could support situations which don't support USM.  
I propose tagging this as an issue for the future, and make the scope wider for a general utility structure, there are many such temp storage buffers in oneDPL I believe which could benefit from such treatment.

Reply:
Actually, now that I'm looking to make the issue I'm remembering #1773 where `__result_and_scratch_storage` was already adapted to be able to provide temporary-only storage, so not much is to be done here. 
I still think it perhaps deserves to be a separate issue to convert this and other temporary storage buffers to use this struct, or perhaps a layer on top of it.
We had agreed to not spend much effort polishing this PR with the intention to improve the set algorithms further.

Reply:
I've created this issue.  https://github.com/oneapi-src/oneDPL/issues/1906
It has some nice potential for a performance improvement across a number of APIs with not too much development effort.  It will require a good amount of correctness and benchmarking to check it though.

Reply:
Thanks!

Reply:
Thank you for the idea.